src/InlineHtmlStripStylesTransformer.ts

```diff
@@ -0,0 +1,260 @@
+const STYLES = 'styles'
+/** Angular component decorator Template property name */
+const TEMPLATE = 'template'
+/** Node require function name */
+const REQUIRE = 'require'
+
+/**
+ * Property names inside the decorator argument to transform
+ */
+const TRANSFORM_PROPS = [TEMPLATE_URL, STYLES, STYLE_URLS]
+
+/**
+ * Transformer ID
+ * @internal
+ */
+export const name = 'angular-component-inline-template-strip-styles'
+
+// increment this each time the code is modified
+/**
+ * Transformer Version
+ * @internal
+ */
+export const version = 1
+
+/**
+ * The factory of hoisting transformer factory
+ * @internal
+ */
+export function factory(cs: ConfigSet) {
+
+  /**
+   * Array Flatten function, as there were problems to make the compiler use
+   * esnext's Array.flat, can be removed in the future.
+   * @param arr Array to be flattened
+   */
+  function flatten<T>(arr: (T | T[] | T[][])[]): T[] {
+    return arr.reduce(
+      (xs: T[], x) => Array.isArray(x) ? xs.concat(flatten(x as T[])) : xs.concat(x),
+      []
+    ) as T[]
+  }
+
+  /**
+   * Our compiler (typescript, or a module with typescript-like interface)
+   */
+  const ts = cs.compilerModule
+
+  /**
+   * Traverses the AST down to the relevant assignments in the decorator
+   * argument and returns them in an array.
+   */
+  function getPropertyAssignmentsToTransform(classDeclaration: ClassDeclaration) {
+    return flatten<PropertyAssignment>(classDeclaration.decorators!
+      .map(dec => dec.expression)
+      .filter(isCallExpression)
+      .map(callExpr => (callExpr as CallExpression).arguments
+        .filter(isObjectLiteralExpression)
+        .map(arg => (arg as ObjectLiteralExpression).properties
+          .filter(isPropertyAssignment)
+          .map(arg => arg as PropertyAssignment)
+          .filter(propAss => isIdentifier(propAss.name))
+          .filter(propAss => TRANSFORM_PROPS.includes((propAss.name as Identifier).text))
+        )
+      )
+    )
+  }
+
+  /**
+   * Clones the node, identifies the properties to transform in the decorator and modifies them.
+   * @param node class declaration with decorators
+   */
+  function transfromDecoratorForJest(node: ClassDeclaration) {
+
+    const mutable = ts.getMutableClone(node)
+    const assignments = getPropertyAssignmentsToTransform(mutable)
+
+    assignments.forEach(assignment => {
+      switch((assignment.name as Identifier).text) {
+        case TEMPLATE_URL:
+          // we can reuse the right-hand-side literal from the assignment
+          let templatePathLiteral = assignment.initializer
+
+          // fix templatePathLiteral if it was a non-relative path
+          if (isStringLiteral(assignment.initializer)) {
+            const templatePathStringLiteral: StringLiteral = assignment.initializer;
+            // match if it starts with ./ or ../ or /
+            if (templatePathStringLiteral.text &&
+              !templatePathStringLiteral.text.match(/^(\.\/|\.\.\/|\/)/)) {
```

Comment:
I haven't tested this improvement yet but by only reading the codes, it's not so generic to me. Because it's not so sure where the users can put their template, it might be ./ , ../../ or some other places within the project. I think it's better to resolve the path within the project or at least within src/app

Reply:
True!
I just checked the old regex mechanism, it would also not support `templateUrl: 'src/page.html'` and transform it to `./src/page.html`.
When the astTransformer hits, it only gets `config-set` injected. Luckily, the ConfigSet object seems to have a method [resolvePath](https://github.com/kulshekhar/ts-jest/blob/984eb8558887c727a6be6eaa02825cb4b882b1bb/src/config/config-set.ts#L700), I will see later if this serves our purpose on checking if `require` will be able to find the file later on.

Reply:
Ok I looked into the issue a bit and saw the ts-compiler itself (which is also available in the transformer) offers many more path-tools than expected:
`toPath, normalizePath, normalizePathAndParts, getDirectoryPath, isRootedDiskPath, isDiskPathRoot, convertToRelativePath, getPathComponents, reducePathComponents, getNormalizedPathComponents, getNormalizedAbsolutePath, getPathFromPathComponents, getRelativePathFromFile, getRelativePathFromDirectory, getRelativePathToDirectoryOrUrl, ensurePathIsNonModuleName, combinePaths, resolvePath, comparePathsCaseSensitive, comparePathsCaseInsensitive, comparePaths, containsPath, extensionFromPath, tryGetExtensionFromPath, getAnyExtensionFromPath, fullTripleSlashReferencePathRegEx, fullTripleSlashAMDReferencePathRegEx, getExternalModuleNameFromPath, getOwnEmitOutputFilePath, getDeclarationEmitOutputFilePath, getSourceFilePathInNewDir, getOutputPathsFor, updateMissingFilePathsWatch, getDefaultLibFilePath`

But it feel like it would be an own task to figure out which methods help us to figure out if the file path is relative to the transformed file or relative to the project root using any path-definition inside `tsconfig.json`.

I quickly checked `resolvePath`, but that always appends the file to the project root.

Would you be ok with opening a new issue and for now keep the same behavior as the regex did before?

Reply:
Of course opening another PR for it is always fine by us👌 The only problem is we will need to create a temporary branch for this PR to merge in instead of master. Because master branch at the moment doesn't encounter this issue therefore we want to keep the state like it is right now. Until we solve the issue with this resolve path we can merge the branch of this PR into master. Do you agree @thymikee ?

Reply:
hmm ok. I hope we make the users happy with this breaking change before they find out this hidden bug :)

Reply:
#### Quick Intro AST and the transformer
(Ignore if you know this already)

If you play with a [ast-viewer](https://ts-ast-viewer.com/) by putting in a simple Angular Component you can get familiar with the AST that TSC will parse form the code. It is a big tree of nested nodes.

Every AST node will have a `kind`-property, which is a SyntaxKind (all of them listed in [typescript.d.ts](https://github.com/Microsoft/TypeScript/blob/8ddb2b61d2c996ab442b282ab9b36195ed697c13/lib/typescript.d.ts#L78)).

For every SyntaxKind there is a `is<KIND>(node)`-method, e. g. `isStringLiteral(node)`. I use these in the transformer, [e. g. here](https://github.com/thymikee/jest-preset-angular/pull/204/files#diff-d8c58610f25910c24023a249bf950abbR137), and can then safely assume the node is of that kind and has the corresponding fields, so there is no need to check against the number 'manually'.

Once called `getMutableClone()` on a node it is possible to modify inner nodes. Nodes can also be programmatically created, as explained [here in the TS docs](https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API#creating-and-printing-a-typescript-ast). We have to do this in the transformer to e. g. [create the `require()`-call](https://github.com/thymikee/jest-preset-angular/pull/204/files#diff-d8c58610f25910c24023a249bf950abbR179).

Being familiar with this enables you to write custom TSLint rules as well.